<a href="https://colab.research.google.com/github/YuliiaUtkina/Git_modul_pyhton/blob/main/hw_3_2_%D0%9E%D1%81%D0%BD%D0%BE%D0%B2%D0%B8_%D1%80%D0%BE%D0%B1%D0%BE%D1%82%D0%B8_%D0%B7_%D0%B4%D0%B0%D0%BD%D0%B8%D0%BC%D0%B8_%D1%83_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Основи роботи з даними у Pandas"

In [6]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [7]:
import pandas as pd
df = pd.read_csv('/content/drive/MyDrive/Data/supermarket_sales.csv')

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [4]:
df.head(4)

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [5]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [6]:
df[615:619]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [12]:
df.columns

Index(['Invoice ID', 'Branch', 'City', 'Customer type', 'Gender',
       'Product line', 'Unit price', 'Quantity', 'Tax 5%', 'Total', 'Date',
       'Time', 'Payment', 'cogs', 'gross margin percentage', 'gross income',
       'Rating'],
      dtype='object')

In [14]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

In [21]:
# ще варіант назви колонок та їх типів
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 17 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Invoice ID               1000 non-null   object 
 1   Branch                   1000 non-null   object 
 2   City                     1000 non-null   object 
 3   Customer type            1000 non-null   object 
 4   Gender                   1000 non-null   object 
 5   Product line             1000 non-null   object 
 6   Unit price               1000 non-null   float64
 7   Quantity                 1000 non-null   int64  
 8   Tax 5%                   1000 non-null   float64
 9   Total                    1000 non-null   float64
 10  Date                     1000 non-null   object 
 11  Time                     1000 non-null   object 
 12  Payment                  1000 non-null   object 
 13  cogs                     1000 non-null   float64
 14  gross margin percentage  

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [41]:
display(df['gross income'].describe(percentiles = [0.01, 0.25, 0.5, 0.75, 0.99]))

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

Це означає, що 75% клієнтів/транзакцій мають "gross income", який менше або рівний значенню 22.45.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [42]:
df.City.value_counts()

Yangon       340
Mandalay     332
Naypyitaw    328
Name: City, dtype: int64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [45]:
df.City.value_counts()/df.City.value_counts().sum()

Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: City, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [80]:
# Середнє значення для міста Yangon
avr_income_yangon = df.query("City == 'Yangon'")['gross income'].mean()

# Середнє значення для міста Mandalay
avr_income_mandalay = df.query("City == 'Mandalay'")['gross income'].mean()

# Виведення результатів
display((avr_income_yangon, avr_income_mandalay))



(14.874001470588235, 15.232024096385542)

У місті Yangon хоч і кількість інвойсів більша, але при цьому gross income менша. Можливі причини (гіпотези):
1. Середній чек/платіж на одного покупця у Yangon менший, ніж у Mandalay.
Якщо гіпотеза про вищий середній чек підтвердиться, можна запровадити програми лояльності чи пропозиції, що стимулюють більші покупки у місті з меншим "gross income".

2. У Yangon купують товари з меншим цінником, ніж у Mandalay. Проаналізувати товари у чеку.
У місті з вищим "gross income" може бути вищий рівень середньомісячного доходу населення, що стимулює більше витрат на продукти та послуги.
Можна порівняти статистичні дані про середній рівень доходу населення для кожного міста, використовуючи, наприклад, офіційні документи чи дослідження.
2.1 Чи на полицях представлений асортимент товарів дешевше за ціною і тому середній чек нижче (особливість регіону або демпінг цін з метою конкурувати з іншими супермаркетами).
Проаналізувати ассортимент товарів у магазинах.

3. Проаналізувати вік покупців у двох магазинах: можливо, у Yangon переважують покупці з категорії студенти/пенсіонери, які вважаються менш платоспроможними, ніж "працююча" когорта.
І тоді варто переглянути цінову політику (у менший бік і асортимент товару (розширити полиці вибором дорожчих або дешевих товарів).

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [15]:
result = df['gross margin percentage'].nunique()
if result == 1:
  print("У всіх всіх інвойсах однакова маржа")
else:
  print("Маржа відрізняється для різних інвойсів")


У всіх всіх інвойсах однакова маржа


Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [17]:
# Фільтруємо дані для міста Yangon,рахуємо кількість інвойсів для кожної категорії товарів
# Обчислюємо відсотковий розподіл для міста Yangon

yangon_percentage = (df[df['City'] == 'Yangon']['Product line'].value_counts() / yangon_invoices.sum()) * 100
print(yangon_percentage)

Home and lifestyle        19.117647
Electronic accessories    17.647059
Sports and travel         17.352941
Food and beverages        17.058824
Fashion accessories       15.000000
Health and beauty         13.823529
Name: Product line, dtype: float64


In [16]:
# Фільтруємо дані для міста Mandalay, рахуємо кількість інвойсів для кожної категорії товарів
# Обчислюємо відсотковий розподіл для міста Mandalay

mandalay_percentage = (df[df['City'] == 'Mandalay']['Product line'].value_counts() / mandalay_invoices.sum()) * 100
print(mandalay_percentage)

Fashion accessories       18.674699
Sports and travel         18.674699
Electronic accessories    16.566265
Health and beauty         15.963855
Food and beverages        15.060241
Home and lifestyle        15.060241
Name: Product line, dtype: float64


Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)